<a href="https://colab.research.google.com/github/seldoncode/Python_CoderDojo/blob/main/Python_CoderDojo12.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Bucles ```for``` anidados

In [ ]:
for i in range(1,4):           # i toma los valores de 1 a 3
  print("~"*10)
  for j in range(5):           # j toma los valores de 0 a 4
    print(f"i={i}, j={j}.")

~~~~~~~~~~
i=1, j=0.
i=1, j=1.
i=1, j=2.
i=1, j=3.
i=1, j=4.
~~~~~~~~~~
i=2, j=0.
i=2, j=1.
i=2, j=2.
i=2, j=3.
i=2, j=4.
~~~~~~~~~~
i=3, j=0.
i=3, j=1.
i=3, j=2.
i=3, j=3.
i=3, j=4.


### Crear una baraja española utilizando bucles anidados
Disponemos de una lista con los tantos y otra con los palos.
* tantos = ['A','2','3','4','5','6','7','S','C','R']
* palos = ['oros','copas','espadas','bastos']

#### Solución 1

In [ ]:
tantos = ['A','2','3','4','5','6','7','S','C','R']
palos = ['oros','copas','espadas','bastos']
for tanto in tantos:
  for palo in palos:
    carta = f"{tanto} de {palo}"
    print(carta)
  print("="*15)

A de oros
A de copas
A de espadas
A de bastos
2 de oros
2 de copas
2 de espadas
2 de bastos
3 de oros
3 de copas
3 de espadas
3 de bastos
4 de oros
4 de copas
4 de espadas
4 de bastos
5 de oros
5 de copas
5 de espadas
5 de bastos
6 de oros
6 de copas
6 de espadas
6 de bastos
7 de oros
7 de copas
7 de espadas
7 de bastos
S de oros
S de copas
S de espadas
S de bastos
C de oros
C de copas
C de espadas
C de bastos
R de oros
R de copas
R de espadas
R de bastos


#### Solución 2
Creando la lista baraja que contiene todas las cartas e imprimiendo luego esa lista.

In [ ]:
tantos = ['A','2','3','4','5','6','7','S','C','R']
palos = ['oros','copas','espadas','bastos']
baraja = []
for tanto in tantos:
  for palo in palos:
    carta = f"{tanto} de {palo}"
    baraja.append(carta)
print(baraja)

['A de oros', 'A de copas', 'A de espadas', 'A de bastos', '2 de oros', '2 de copas', '2 de espadas', '2 de bastos', '3 de oros', '3 de copas', '3 de espadas', '3 de bastos', '4 de oros', '4 de copas', '4 de espadas', '4 de bastos', '5 de oros', '5 de copas', '5 de espadas', '5 de bastos', '6 de oros', '6 de copas', '6 de espadas', '6 de bastos', '7 de oros', '7 de copas', '7 de espadas', '7 de bastos', 'S de oros', 'S de copas', 'S de espadas', 'S de bastos', 'C de oros', 'C de copas', 'C de espadas', 'C de bastos', 'R de oros', 'R de copas', 'R de espadas', 'R de bastos']


#### Solución 3

In [ ]:
tantos = ['A','2','3','4','5','6','7','S','C','R']
palos = ['oros','copas','espadas','bastos']
baraja = []
for t in tantos:
  for p in palos:
    carta = f"{t} de {p}"
    baraja.append(carta)

for i in range(0,40,4):
  print(f"{baraja[i]}, {baraja[i+1]}, {baraja[i+2]}, {baraja[i+3]}.")

A de oros, A de copas, A de espadas, A de bastos.
2 de oros, 2 de copas, 2 de espadas, 2 de bastos.
3 de oros, 3 de copas, 3 de espadas, 3 de bastos.
4 de oros, 4 de copas, 4 de espadas, 4 de bastos.
5 de oros, 5 de copas, 5 de espadas, 5 de bastos.
6 de oros, 6 de copas, 6 de espadas, 6 de bastos.
7 de oros, 7 de copas, 7 de espadas, 7 de bastos.
S de oros, S de copas, S de espadas, S de bastos.
C de oros, C de copas, C de espadas, C de bastos.
R de oros, R de copas, R de espadas, R de bastos.


#### Solución 4

In [ ]:
tantos = ['A','2','3','4','5','6','7','S','C','R']
palos = ['oros','copas','espadas','bastos']
baraja = []
for t in tantos:
  for p in palos:
    carta = f"{t} de {p}"
    baraja.append(carta)

for i in range(0,40,4):
  for j in range(4):
    print(f" {baraja[i+j]:16} ", end="")
  print()

 A de oros         A de copas        A de espadas      A de bastos      
 2 de oros         2 de copas        2 de espadas      2 de bastos      
 3 de oros         3 de copas        3 de espadas      3 de bastos      
 4 de oros         4 de copas        4 de espadas      4 de bastos      
 5 de oros         5 de copas        5 de espadas      5 de bastos      
 6 de oros         6 de copas        6 de espadas      6 de bastos      
 7 de oros         7 de copas        7 de espadas      7 de bastos      
 S de oros         S de copas        S de espadas      S de bastos      
 C de oros         C de copas        C de espadas      C de bastos      
 R de oros         R de copas        R de espadas      R de bastos      


#### Solución 5

In [ ]:
tantos = ['A','2','3','4','5','6','7','S','C','R']
palos = ['oros','copas','espadas','bastos']
baraja = []
for t in tantos:
  for p in palos:
    carta = f"{t} de {p}"
    baraja.append(carta)

for p in palos:
  for t in tantos:
    print(f"{t} de {p}")
  print("*"*15)

A de oros
2 de oros
3 de oros
4 de oros
5 de oros
6 de oros
7 de oros
S de oros
C de oros
R de oros
***************
A de copas
2 de copas
3 de copas
4 de copas
5 de copas
6 de copas
7 de copas
S de copas
C de copas
R de copas
***************
A de espadas
2 de espadas
3 de espadas
4 de espadas
5 de espadas
6 de espadas
7 de espadas
S de espadas
C de espadas
R de espadas
***************
A de bastos
2 de bastos
3 de bastos
4 de bastos
5 de bastos
6 de bastos
7 de bastos
S de bastos
C de bastos
R de bastos
***************


### Reto 12.1. Baraja inglesa
Imprime las cartas de una baraja inglesa, donde se imprimar agrupados los tantos de cada palo.  
Primero se imprimirían todos los tantos de picas, luego de corazones, etc:
* ♠ → Picas
* ♥ → Corazones
* ♦ → Rombos
* ♣ → Tréboles

Los tantos son:
* A → As
*	2
*	3
*	4
*	5
*	6
*	7
*	8
*	9
*	10
*	J → Jota/Sota/Jack
*	Q → Reina/Dama
*	K → Rey

### Tablas de multiplicar
Programar las tablas de multiplicar del 2 al 9 para que salgan en diferentes columnas y no una debajo de otra.

In [ ]:
for i in range(1,11):
  for j in range(2,10):
    print(f"{i:2} x {j:1} = {i*j:2}", end=" |")
  print()

 1 x 2 =  2 | 1 x 3 =  3 | 1 x 4 =  4 | 1 x 5 =  5 | 1 x 6 =  6 | 1 x 7 =  7 | 1 x 8 =  8 | 1 x 9 =  9 |
 2 x 2 =  4 | 2 x 3 =  6 | 2 x 4 =  8 | 2 x 5 = 10 | 2 x 6 = 12 | 2 x 7 = 14 | 2 x 8 = 16 | 2 x 9 = 18 |
 3 x 2 =  6 | 3 x 3 =  9 | 3 x 4 = 12 | 3 x 5 = 15 | 3 x 6 = 18 | 3 x 7 = 21 | 3 x 8 = 24 | 3 x 9 = 27 |
 4 x 2 =  8 | 4 x 3 = 12 | 4 x 4 = 16 | 4 x 5 = 20 | 4 x 6 = 24 | 4 x 7 = 28 | 4 x 8 = 32 | 4 x 9 = 36 |
 5 x 2 = 10 | 5 x 3 = 15 | 5 x 4 = 20 | 5 x 5 = 25 | 5 x 6 = 30 | 5 x 7 = 35 | 5 x 8 = 40 | 5 x 9 = 45 |
 6 x 2 = 12 | 6 x 3 = 18 | 6 x 4 = 24 | 6 x 5 = 30 | 6 x 6 = 36 | 6 x 7 = 42 | 6 x 8 = 48 | 6 x 9 = 54 |
 7 x 2 = 14 | 7 x 3 = 21 | 7 x 4 = 28 | 7 x 5 = 35 | 7 x 6 = 42 | 7 x 7 = 49 | 7 x 8 = 56 | 7 x 9 = 63 |
 8 x 2 = 16 | 8 x 3 = 24 | 8 x 4 = 32 | 8 x 5 = 40 | 8 x 6 = 48 | 8 x 7 = 56 | 8 x 8 = 64 | 8 x 9 = 72 |
 9 x 2 = 18 | 9 x 3 = 27 | 9 x 4 = 36 | 9 x 5 = 45 | 9 x 6 = 54 | 9 x 7 = 63 | 9 x 8 = 72 | 9 x 9 = 81 |
10 x 2 = 20 |10 x 3 = 30 |10 x 4 = 40 |10 x 5 = 50 |10 

### Fuerza bruta: combinatoria
* Tenemos un candado de tres ruedas.
* Cada rueda tiene los números del 0 al 9
* Se nos ha olvidado la contraseña
* Probar todas las combinaciones posibles
* Mostrarlas en pantalla

#### Solución
Hablamos de combinaciones pero en realidad, en este caso, se tratan de las variaciones con repetición de 10 elementos tomados de 3 en 3.

[http://www.apuntesmareaverde.org.es/grupos/mat/4B/14_Combinatoria_4B.pdf](http://www.apuntesmareaverde.org.es/grupos/mat/4B/14_Combinatoria_4B.pdf)

El número de posibilidades es $$VR(n,m) = n^m = 10^3=1.000$$


In [ ]:
for i in range(10):
  for j in range(10):
    for k in range(10):
      print(str(i) + str(j) + str(k))  # concatenando los strings

000
001
002
003
004
005
006
007
008
009
010
011
012
013
014
015
016
017
018
019
020
021
022
023
024
025
026
027
028
029
030
031
032
033
034
035
036
037
038
039
040
041
042
043
044
045
046
047
048
049
050
051
052
053
054
055
056
057
058
059
060
061
062
063
064
065
066
067
068
069
070
071
072
073
074
075
076
077
078
079
080
081
082
083
084
085
086
087
088
089
090
091
092
093
094
095
096
097
098
099
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249


### Reto 12.1. Candado de cuatro ruedas
* Se nos ha olvidado la clave de un candado de 4 ruedas.
* Cada rueda contiene los números del 0 al 9.
* Generar las posibles combinaciones del candado.
* Mostrar en pantalla todas las claves posibles probando por fuerza bruta.

## Conjunto
Un conjunto (set) es similar a una lista pero:
* va entre llaves, por ejemplo: {1,2,3}
* no acepta reptidos
* si se intentan incluir elemento repteidos los elimina
* una lista se puede convertir en un conjunto con set
* lista = [1,2,3]
* conjunto = set(lista)
* conjunto es {1,2,3}

In [ ]:
lista = [1,2,3]
conjunto = set(lista)  # convertimos la lista en un conjunto
conjunto

{1, 2, 3}

In [ ]:
lista = [1,2,3,2,2,3]
conjunto = set(lista)  # elimina repetidos
conjunto

{1, 2, 3}

### Cinco amigos hacen una carrera
* A = Ana
* E = Edu
* I = Isa
* O = Oscar
* U = Unai

Queremos saber todas las posibles combinaciones en las que pueden llegar a la meta.

Incluir un contador para calcular cuantas posibilidades existen.

#### Solución
Se trata de las permutaciones sin repetición de 5 elementos que es 5 factorial.

$$P(n) = n! = 5! = 5*4*3*2*1 = 120$$

In [ ]:
amigos = "AEIOU"
lista =list(amigos)
contador = 0
for i in lista:
  for j in lista:
    for k in lista:
      for l in lista:
        for m in lista:
          if i!=j and i!=k and i!=l and i!=m and \
            j!=k and j!=l and j!=m and \
            k!=l and k!=m and \
            l!=m:
            orden =[i,j,k,l,m]
            contador += 1
            print(f"{contador:3d} → {orden}")

  1 → ['A', 'E', 'I', 'O', 'U']
  2 → ['A', 'E', 'I', 'U', 'O']
  3 → ['A', 'E', 'O', 'I', 'U']
  4 → ['A', 'E', 'O', 'U', 'I']
  5 → ['A', 'E', 'U', 'I', 'O']
  6 → ['A', 'E', 'U', 'O', 'I']
  7 → ['A', 'I', 'E', 'O', 'U']
  8 → ['A', 'I', 'E', 'U', 'O']
  9 → ['A', 'I', 'O', 'E', 'U']
 10 → ['A', 'I', 'O', 'U', 'E']
 11 → ['A', 'I', 'U', 'E', 'O']
 12 → ['A', 'I', 'U', 'O', 'E']
 13 → ['A', 'O', 'E', 'I', 'U']
 14 → ['A', 'O', 'E', 'U', 'I']
 15 → ['A', 'O', 'I', 'E', 'U']
 16 → ['A', 'O', 'I', 'U', 'E']
 17 → ['A', 'O', 'U', 'E', 'I']
 18 → ['A', 'O', 'U', 'I', 'E']
 19 → ['A', 'U', 'E', 'I', 'O']
 20 → ['A', 'U', 'E', 'O', 'I']
 21 → ['A', 'U', 'I', 'E', 'O']
 22 → ['A', 'U', 'I', 'O', 'E']
 23 → ['A', 'U', 'O', 'E', 'I']
 24 → ['A', 'U', 'O', 'I', 'E']
 25 → ['E', 'A', 'I', 'O', 'U']
 26 → ['E', 'A', 'I', 'U', 'O']
 27 → ['E', 'A', 'O', 'I', 'U']
 28 → ['E', 'A', 'O', 'U', 'I']
 29 → ['E', 'A', 'U', 'I', 'O']
 30 → ['E', 'A', 'U', 'O', 'I']
 31 → ['E', 'I', 'A', 'O', 'U']
 32 → ['

#### Solución 2

In [ ]:
amigos = "AEIOU"
lista =list(amigos)
contador = 0
for i in lista:
  for j in lista:
    for k in lista:
      for l in lista:
        for m in lista:
          orden =[i,j,k,l,m]
          if len(set(orden))==5:  # set convierte la lista en un conjunto que no admite repetidos
            contador += 1
            print(f"{contador:3d} → {orden}")

  1 → ['A', 'E', 'I', 'O', 'U']
  2 → ['A', 'E', 'I', 'U', 'O']
  3 → ['A', 'E', 'O', 'I', 'U']
  4 → ['A', 'E', 'O', 'U', 'I']
  5 → ['A', 'E', 'U', 'I', 'O']
  6 → ['A', 'E', 'U', 'O', 'I']
  7 → ['A', 'I', 'E', 'O', 'U']
  8 → ['A', 'I', 'E', 'U', 'O']
  9 → ['A', 'I', 'O', 'E', 'U']
 10 → ['A', 'I', 'O', 'U', 'E']
 11 → ['A', 'I', 'U', 'E', 'O']
 12 → ['A', 'I', 'U', 'O', 'E']
 13 → ['A', 'O', 'E', 'I', 'U']
 14 → ['A', 'O', 'E', 'U', 'I']
 15 → ['A', 'O', 'I', 'E', 'U']
 16 → ['A', 'O', 'I', 'U', 'E']
 17 → ['A', 'O', 'U', 'E', 'I']
 18 → ['A', 'O', 'U', 'I', 'E']
 19 → ['A', 'U', 'E', 'I', 'O']
 20 → ['A', 'U', 'E', 'O', 'I']
 21 → ['A', 'U', 'I', 'E', 'O']
 22 → ['A', 'U', 'I', 'O', 'E']
 23 → ['A', 'U', 'O', 'E', 'I']
 24 → ['A', 'U', 'O', 'I', 'E']
 25 → ['E', 'A', 'I', 'O', 'U']
 26 → ['E', 'A', 'I', 'U', 'O']
 27 → ['E', 'A', 'O', 'I', 'U']
 28 → ['E', 'A', 'O', 'U', 'I']
 29 → ['E', 'A', 'U', 'I', 'O']
 30 → ['E', 'A', 'U', 'O', 'I']
 31 → ['E', 'I', 'A', 'O', 'U']
 32 → ['

### Dando nombre a 20.000 planetas
* Con el telescopio espacial es posible que se descubran 20.000 planetas
* Una asociación astronómica quiere dar nombre a todos ellos.
* Deseamos formar nombres de 3 sílabas que se puedan formar con 3 consonantes y 5 vocales.
* En el nombre se alterna consonante y vocal
* Será: consonante + vocal + consonante + vocal + consonante + vocal
* No se repite ninguna letra en cada nombre
* Se genera una lista con todas las combinaciones
* Esa lista no se mustra porque es muy larga
* De la lista se mustran aleatoriamente 20 nombres.

#### Solución
* El orden si importa, ya que no es lo mismo 'cosati' que 'ticosa'.
* Se toman algunos elementos ya que de las 10 consonantes disponibles se toman 3, y de las 5 vocales vamos a tomar 3.
* Sin repetición, ya que no se puede repetir ninguna consonante ni ninguna vocal.
* Se trata de VARIACIONES sin repetición.

$$V(n,m) = \frac{m!}{(m-n)!}$$

* Aplicando la fórmula a las consonantes tenemos:
$$\frac{10!}{(10-3)!} = 10*9*8 =720$$

* Aplicando la fórmula a las vocales tenemos:
$$\frac{5!}{(5-3)!} = 5*4*3 =60$$

* Total = Consonantes * Vocales
* Total = 710 * 60 = 43.200

* Como son 20.000 planetas nos sobran combinaciones de nombres para atender a todos.

In [ ]:
import random
consonantes = "bcdflnprst"
vocales = "aeiou"
nombres = []  # lista que recoge los nombres que se formen
# seis bucles for ya que son nombres de 6 letras
for c1 in consonantes:
  for v1 in vocales:
    for c2 in consonantes:
      for v2 in vocales:
        for c3 in consonantes:
          for v3 in vocales:
            if c1!=c2 and c1!=c3 and c2!=c3 and \
               v1!=v2 and v1!=v3 and v2!=v3:           # si todas las consonantes son distintas y todas las vocales son distintas
               nombre = c1+v1+c2+v2+c3+v3
               nombres.append(nombre.capitalize())
               #print(nombre)
print(f"Se han generado {len(nombres)} nombres.")
print("De entre ellos elegimos 20 nombres.")
random.choices(nombres, k=20)

Se han generado 43200 nombres.
De entre ellos elegimos 20 nombres.


['Sofeta',
 'Pufice',
 'Cenisa',
 'Nalebi',
 'Sedubo',
 'Poreni',
 'Nerosu',
 'Cibote',
 'Ducole',
 'Leconi',
 'Bopesu',
 'Topebu',
 'Fesanu',
 'Rotasu',
 'Delino',
 'Fepobu',
 'Bafeci',
 'Cinofu',
 'Funola',
 'Dasibo']

### Lotería
* En una lotería hay un bombo que tiene 10 bolas, con los números del 0 al 9.
* En un sorteo se sacan 5 bolas
* Las bolsas que se sacan no se vuleven a introducir
* Queremos conocer todas las apustas posibles, esto es
* queremos conocer todas las combinaciones posibles que se pueden dar en esa lotería
* Incluiremos un contador para comprobar el número de combinaciones obtenidas.

#### Solución 1
* Son combinaciones sin repetición de 10 elementos tomados de 5 en 5.

$$C(10,5)=\frac{m!}{n!*(m-n)!}=\frac{10!}{5!*(10-5)!}=\frac{10*9*8*7*6}{5*4*3*2*1}=\frac{30240}{120}=252$$

In [ ]:
numeros = '9876543210'
bolas = list(numeros)
bolas.sort()
bolas
contador = 0
combinaciones = []
for i in range(10):
  for j in range(10):
    for k in range(10):
      for l in range(10):
        for m in range(10):
          n = str(i) + str(j) + str(k) + str(l) + str(m)
          n=sorted(n)
          if len(set(n))==5 and n not in combinaciones:
            combinaciones.append(n)
            contador += 1
print(f"Existen {contador} combinaciones")
print(combinaciones)

Existen 252 combinaciones
[['0', '1', '2', '3', '4'], ['0', '1', '2', '3', '5'], ['0', '1', '2', '3', '6'], ['0', '1', '2', '3', '7'], ['0', '1', '2', '3', '8'], ['0', '1', '2', '3', '9'], ['0', '1', '2', '4', '5'], ['0', '1', '2', '4', '6'], ['0', '1', '2', '4', '7'], ['0', '1', '2', '4', '8'], ['0', '1', '2', '4', '9'], ['0', '1', '2', '5', '6'], ['0', '1', '2', '5', '7'], ['0', '1', '2', '5', '8'], ['0', '1', '2', '5', '9'], ['0', '1', '2', '6', '7'], ['0', '1', '2', '6', '8'], ['0', '1', '2', '6', '9'], ['0', '1', '2', '7', '8'], ['0', '1', '2', '7', '9'], ['0', '1', '2', '8', '9'], ['0', '1', '3', '4', '5'], ['0', '1', '3', '4', '6'], ['0', '1', '3', '4', '7'], ['0', '1', '3', '4', '8'], ['0', '1', '3', '4', '9'], ['0', '1', '3', '5', '6'], ['0', '1', '3', '5', '7'], ['0', '1', '3', '5', '8'], ['0', '1', '3', '5', '9'], ['0', '1', '3', '6', '7'], ['0', '1', '3', '6', '8'], ['0', '1', '3', '6', '9'], ['0', '1', '3', '7', '8'], ['0', '1', '3', '7', '9'], ['0', '1', '3', '8', '9'], [

#### Solución 2

In [ ]:
combinaciones = []
contador = 0
for i in range(10):
  for j in range(10):
    for k in range(10):
      for l in range(10):
        for m in range(10):
          if i!=j and i!=k and i!=l and i!=m and \
             j!=k and j!=l and j!=m and \
             k!=l and k!=m and \
             l!=m:
             comb = [i,j,k,l,m]
             comb.sort()
             if comb not in combinaciones:
               combinaciones.append(comb)
               contador += 1
               print(f"{contador:3d} {comb}")


  1 [0, 1, 2, 3, 4]
  2 [0, 1, 2, 3, 5]
  3 [0, 1, 2, 3, 6]
  4 [0, 1, 2, 3, 7]
  5 [0, 1, 2, 3, 8]
  6 [0, 1, 2, 3, 9]
  7 [0, 1, 2, 4, 5]
  8 [0, 1, 2, 4, 6]
  9 [0, 1, 2, 4, 7]
 10 [0, 1, 2, 4, 8]
 11 [0, 1, 2, 4, 9]
 12 [0, 1, 2, 5, 6]
 13 [0, 1, 2, 5, 7]
 14 [0, 1, 2, 5, 8]
 15 [0, 1, 2, 5, 9]
 16 [0, 1, 2, 6, 7]
 17 [0, 1, 2, 6, 8]
 18 [0, 1, 2, 6, 9]
 19 [0, 1, 2, 7, 8]
 20 [0, 1, 2, 7, 9]
 21 [0, 1, 2, 8, 9]
 22 [0, 1, 3, 4, 5]
 23 [0, 1, 3, 4, 6]
 24 [0, 1, 3, 4, 7]
 25 [0, 1, 3, 4, 8]
 26 [0, 1, 3, 4, 9]
 27 [0, 1, 3, 5, 6]
 28 [0, 1, 3, 5, 7]
 29 [0, 1, 3, 5, 8]
 30 [0, 1, 3, 5, 9]
 31 [0, 1, 3, 6, 7]
 32 [0, 1, 3, 6, 8]
 33 [0, 1, 3, 6, 9]
 34 [0, 1, 3, 7, 8]
 35 [0, 1, 3, 7, 9]
 36 [0, 1, 3, 8, 9]
 37 [0, 1, 4, 5, 6]
 38 [0, 1, 4, 5, 7]
 39 [0, 1, 4, 5, 8]
 40 [0, 1, 4, 5, 9]
 41 [0, 1, 4, 6, 7]
 42 [0, 1, 4, 6, 8]
 43 [0, 1, 4, 6, 9]
 44 [0, 1, 4, 7, 8]
 45 [0, 1, 4, 7, 9]
 46 [0, 1, 4, 8, 9]
 47 [0, 1, 5, 6, 7]
 48 [0, 1, 5, 6, 8]
 49 [0, 1, 5, 6, 9]
 50 [0, 1, 5, 7, 8]


## Módulo ```itertools```
Este módulo es útil en el campo de la combinatoria.

In [ ]:
import itertools
dir(itertools)

['__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_grouper',
 '_tee',
 '_tee_dataobject',
 'accumulate',
 'chain',
 'combinations',
 'combinations_with_replacement',
 'compress',
 'count',
 'cycle',
 'dropwhile',
 'filterfalse',
 'groupby',
 'islice',
 'permutations',
 'product',
 'repeat',
 'starmap',
 'takewhile',
 'tee',
 'zip_longest']

In [ ]:
print("El número de combinaciones es:", len(list(itertools.combinations(range(10),5))))
print()
list(itertools.combinations(range(10),5))

El número de combinaciones es: 252



[(0, 1, 2, 3, 4),
 (0, 1, 2, 3, 5),
 (0, 1, 2, 3, 6),
 (0, 1, 2, 3, 7),
 (0, 1, 2, 3, 8),
 (0, 1, 2, 3, 9),
 (0, 1, 2, 4, 5),
 (0, 1, 2, 4, 6),
 (0, 1, 2, 4, 7),
 (0, 1, 2, 4, 8),
 (0, 1, 2, 4, 9),
 (0, 1, 2, 5, 6),
 (0, 1, 2, 5, 7),
 (0, 1, 2, 5, 8),
 (0, 1, 2, 5, 9),
 (0, 1, 2, 6, 7),
 (0, 1, 2, 6, 8),
 (0, 1, 2, 6, 9),
 (0, 1, 2, 7, 8),
 (0, 1, 2, 7, 9),
 (0, 1, 2, 8, 9),
 (0, 1, 3, 4, 5),
 (0, 1, 3, 4, 6),
 (0, 1, 3, 4, 7),
 (0, 1, 3, 4, 8),
 (0, 1, 3, 4, 9),
 (0, 1, 3, 5, 6),
 (0, 1, 3, 5, 7),
 (0, 1, 3, 5, 8),
 (0, 1, 3, 5, 9),
 (0, 1, 3, 6, 7),
 (0, 1, 3, 6, 8),
 (0, 1, 3, 6, 9),
 (0, 1, 3, 7, 8),
 (0, 1, 3, 7, 9),
 (0, 1, 3, 8, 9),
 (0, 1, 4, 5, 6),
 (0, 1, 4, 5, 7),
 (0, 1, 4, 5, 8),
 (0, 1, 4, 5, 9),
 (0, 1, 4, 6, 7),
 (0, 1, 4, 6, 8),
 (0, 1, 4, 6, 9),
 (0, 1, 4, 7, 8),
 (0, 1, 4, 7, 9),
 (0, 1, 4, 8, 9),
 (0, 1, 5, 6, 7),
 (0, 1, 5, 6, 8),
 (0, 1, 5, 6, 9),
 (0, 1, 5, 7, 8),
 (0, 1, 5, 7, 9),
 (0, 1, 5, 8, 9),
 (0, 1, 6, 7, 8),
 (0, 1, 6, 7, 9),
 (0, 1, 6, 8, 9),
 (0, 1, 7,

### Adivinar una contraseña por fuerza bruta
* Adivinar una contraseña de 4 letras minúsculas por fuerza bruta.
* Medir el tiempo que se tarda en hacer todas las combinaciones.
* alfabeto = "abcdefghijklmnñopqrstuvwxyz"
* En español son 27 letras

In [ ]:
import time
clave = input("Dime una clave de 4 letras minúsculas: ").lower()
alfabeto = "abcdefghijklmnñopqrstuvwxyz"
inicio = time.time()
for c1 in alfabeto:
  for c2 in alfabeto:
    for c3 in alfabeto:
      for c4 in alfabeto:
        intento = c1+c2+c3+c4
        if intento == clave:
          print("Clave encontrada:", intento)
final = time.time()
print("Tiempo empleado:", final-inicio)


Dime una clave de 4 letras minúsculas: Jose
Clave encontrada: jose
Tiempo empleado: 0.18987131118774414


### Contraseña de 5 letras minúsculas
* Mismo caso que el anterior pero con 5 letras minúsculas.
* Observe que existe un lapso de tiempo entre que encuentra la clave y termina de ejecutar todos los bucles for anidados.

In [ ]:
import time
clave = input("Dime una clave de 5 letras minúsculas: ").lower()
alfabeto = "abcdefghijklmnñopqrstuvwxyz"
inicio = time.time()
for c1 in alfabeto:
  for c2 in alfabeto:
    for c3 in alfabeto:
      for c4 in alfabeto:
        for c5 in alfabeto:
          intento = c1+c2+c3+c4+c5
          if intento == clave:
            print("Clave encontrada:", intento)
final = time.time()
print(f"Tiempo empleado: {round(final-inicio)} segundos.")

Dime una clave de 5 letras minúsculas: Anita
Clave encontrada: anita
Tiempo empleado: 6 segundos.


### Contraseña de 5 caracteres con mayúsculas, minúsculas y números
Mismo caso que el anterior pero con 5 caracteres que pueden ser letras minúsculas, mayúsculas o números.

In [ ]:
import time
clave = input("Dime una clave de 5 caracteres (minúsculas, mayúsculas, números): ")
alfabeto = "abcdefghijklmnñopqrstuvwxyzABCDEFGHIJKLMNÑOPQRSTUVWXYZ0123456789"
inicio = time.time()
for c1 in alfabeto:
  for c2 in alfabeto:
    for c3 in alfabeto:
      for c4 in alfabeto:
        for c5 in alfabeto:
          intento = c1+c2+c3+c4+c5
          if intento == clave:
            print("Clave encontrada:", intento)
final = time.time()
print(f"Tiempo empleado aproximado: {round((final-inicio)/60,1)} minutos.")

Dime una clave de 5 caracteres (minúsculas, mayúsculas, números): Ana12
Clave encontrada: Ana12
Tiempo empleado aproximado: 7.5 minutos.


### Contraseña de hasta 5 caracteres
* En base al caso anterior
* Ahora la contraseña introducida puede ser de 1 o 2 o 3 o 4 o 5 caracteres
* Conseguir que cuando se detecte la contraseña no sea necesario esperar hasta que los bucles anidados terminen.

In [ ]:
import time
clave = input("Dime una clave de hasta 5 caracteres (minúsculas, mayúsculas, números): ")
alfabeto = " abcdefghijklmnñopqrstuvwxyzABCDEFGHIJKLMNÑOPQRSTUVWXYZ0123456789" # incluimos un espacio
intento = None
inicio = time.time()
for c1 in alfabeto:
  if intento != clave:                  # añadiendo condicional
    for c2 in alfabeto:
      if intento != clave:              # añadiendo condicional
        for c3 in alfabeto:
          if intento != clave:          # añadiendo condicional
            for c4 in alfabeto:
              if intento != clave:      # añadiendo condicional
                for c5 in alfabeto:
                  cc = c1+c2+c3+c4+c5
                  intento = cc.strip()  # eliminamos los espacios iniciales y finales, así analizamos las claves de menos de 5
                  if intento == clave:
                    print("Clave encontrada:", intento)
                  break
              else: break
          else: break
      else: break
  else: break
final = time.time()
print(f"Tiempo empleado aproximado: {round((final-inicio)/60,1)} minutos.")

Dime una clave de hasta 5 caracteres (minúsculas, mayúsculas, números): Zz9a
Clave encontrada: Zz9a
Tiempo empleado aproximado: 0.2 minutos.


### Otro método para salir de bucles ```for``` anidados

#### Caso 1.A.
* Primero veamos el tiempo parcial y total
* El tiempo parcial hasta fin1 que es cuando se cumple la condición es una fracción del tiempo total.
* El tiempo total es mayor porque se tienen que cumplir todos los bucles ```for``` hasta el final de sus iteraciones.

In [ ]:
import time
inicio = time.time()
for a in range(10000):
  for b in range(10000):
    if a==b==2000:
      print("Todos iguales al valor establecido.")
      fin1 = time.time()

fin2 = time.time()
print(fin1-inicio,"Tiempo hasta que se cumple el condicional")
print(fin2-inicio, "Tiempo total")

Todos iguales al valor establecido.
1.6832444667816162 Tiempo hasta que se cumple el condicional
8.337207555770874 Tiempo total


#### Caso 2.A.
* Ahora vemos que el tiempo total y parcial coinciden
* El motivo es que una vez que se cumple la condición salimos completamente de todos los bucles, sin esperar a que se cumplan todas las iteraciones.

In [ ]:
import time
inicio = time.time()
for a in range(10000):
  for b in range(10000):
    if a==b==2000:
      print("Todos iguales al valor establecido.")
      fin1 = time.time()
      break
  else: continue
  break

fin2 = time.time()
print(fin1-inicio,"Tiempo hasta que se cumple el condicional")
print(fin2-inicio, "Tiempo total")

Todos iguales al valor establecido.
1.679574966430664 Tiempo hasta que se cumple el condicional
1.6796507835388184 Tiempo total


#### Caso 1.B.
Veamos ahora los tiempos total y parcial en un caso con 8 bucles ```for``` anidados.

In [ ]:
import time
inicio = time.time()
for i in range(10):
  for j in range(10):
    for k in range(10):
      for l in range(10):
        for m in range(10):
          for n in range(10):
            for o in range(10):
              for p in range(10):
                if i==j==k==l==m==n==o==p==2:
                  print("Todos iguales al valor establecido.")
                  fin1 = time.time()

fin2 = time.time()
print(fin1-inicio,"Tiempo hasta que se cumple el condicional")
print(fin2-inicio, "Tiempo total")

Todos iguales al valor establecido.
2.370201826095581 Tiempo hasta que se cumple el condicional
10.610024690628052 Tiempo total


#### Caso 2.B.
* En el caso de los 8 bucles anidados conseguimos salir inmediatamente en el momento en el que se cumple la condición sin esperar a que todas las iteraciones terminen.
* La gran diferencia de este caso con el caso 2.A. es que aquí, gracias al ```continue``` no necesitamos ir poniendo condicionales en cada bucle.

In [ ]:
import time
inicio = time.time()
for i in range(10):
  for j in range(10):
    for k in range(10):
      for l in range(10):
        for m in range(10):
          for n in range(10):
            for o in range(10):
              for p in range(10):
                if i==j==k==l==m==n==o==p==2:
                  print("Todos iguales al valor establecido.")
                  fin1 = time.time()
                  break
              else: continue
              break
            else: continue
            break
          else: continue
          break
        else: continue
        break
      else: continue
      break
    else: continue
    break
  else: continue
  break

fin2 = time.time()
print(fin1-inicio,"Tiempo hasta que se cumple el condicional")
print(fin2-inicio, "Tiempo total")

Todos iguales al valor establecido.
2.3656673431396484 Tiempo hasta que se cumple el condicional
2.36576247215271 Tiempo total


### Nota sobre cómo salir eficientemente de bucles anidados
Aún no hemos visto funciones, pero realmente la forma más eficiente para salir de bucles anidados es tener nuestro código dentro de una función, de forma que cuando se cumpla la condición y queramos salir, sin esperar a que se ejecuten todas las iteraciones, lo mejor es poner un ```return``` que finaliza instantáneamente la función.

In [ ]:
import time

def funcion():
  inicio = time.time()
  for a in range(10000):
    for b in range(10000):
      if a==b==2000:
        print("Todos iguales al valor establecido.")
        fin = time.time()
        print(fin-inicio, "Tiempo total")
        return
      
funcion()

Todos iguales al valor establecido.
1.1493608951568604 Tiempo total


# Range con salto
El tercer argumento de range nos indica el salto o paso (step).

In [ ]:
for i in range(1,11,2):     # va de 2 en 2
  print(i)

1
3
5
7
9


In [ ]:
for i in range(12,0,-2):
  print(i)

12
10
8
6
4
2


## Convertir un range en una lista y memoria ocupada


In [ ]:
rango10 = range(10)         # es un objeto de tipo range
rango10

range(0, 10)

In [ ]:
lista10 = list(range(10))   # es un objeto de tipo lista
lista10

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

### Los rangos ocupan menos memoria que las listas

In [ ]:
import sys
bytes_rango10 = sys.getsizeof(rango10)
bytes_lista10 = sys.getsizeof(lista10)
print(bytes_rango10)                        #  48 bytes
print(bytes_lista10)                        # 208 bytes

48
208


In [ ]:
import sys
rango10000 = range(10000) 
lista10000 = list(range(10000))
bytes_rango10000 = sys.getsizeof(rango10000)
bytes_lista10000 = sys.getsizeof(lista10000)
print(bytes_rango10000)                        #    48 bytes, el rango sigue ocupando lo mismo
print(bytes_lista10000)                        # 90120 bytes, la lista ocupa mucho más

48
90120


### Listar 100 números
* Listar los números del 1 al 100 en 10 columnas y 10 filas.
* La primera fila con los números del 1 al 10
* La última fila con los números del 91 al 100.

#### Solución 1

In [ ]:
for i in range(0,10):
  for j in range(1,11):
    print(f"{i*10+j:4}", end="")
  print()

   1   2   3   4   5   6   7   8   9  10
  11  12  13  14  15  16  17  18  19  20
  21  22  23  24  25  26  27  28  29  30
  31  32  33  34  35  36  37  38  39  40
  41  42  43  44  45  46  47  48  49  50
  51  52  53  54  55  56  57  58  59  60
  61  62  63  64  65  66  67  68  69  70
  71  72  73  74  75  76  77  78  79  80
  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100


#### Solución 2
Usando el salto en el rango.

In [ ]:
for i in range(0,91,10):
  for j in range(1,11):
    print(f"{i+j:4}", end="")
  print()

   1   2   3   4   5   6   7   8   9  10
  11  12  13  14  15  16  17  18  19  20
  21  22  23  24  25  26  27  28  29  30
  31  32  33  34  35  36  37  38  39  40
  41  42  43  44  45  46  47  48  49  50
  51  52  53  54  55  56  57  58  59  60
  61  62  63  64  65  66  67  68  69  70
  71  72  73  74  75  76  77  78  79  80
  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100


### Listar 100 números en orden inverso
* Listar los números del 100 al 1 en 10 columnas y 10 filas.
* La primera fila con los números del 100 al 91
* La última fila con los números del 10 al 1.

#### Solución 1

In [ ]:
for i in range(0,10):
  for j in range(1,11):
    print(f"{101-i*10-j:4}", end="")
  print()

 100  99  98  97  96  95  94  93  92  91
  90  89  88  87  86  85  84  83  82  81
  80  79  78  77  76  75  74  73  72  71
  70  69  68  67  66  65  64  63  62  61
  60  59  58  57  56  55  54  53  52  51
  50  49  48  47  46  45  44  43  42  41
  40  39  38  37  36  35  34  33  32  31
  30  29  28  27  26  25  24  23  22  21
  20  19  18  17  16  15  14  13  12  11
  10   9   8   7   6   5   4   3   2   1


#### Solución 2
Usando el salto en el rango.

In [ ]:
for i in range(90,-1,-10):
  for j in range(10,0,-1):
    print(f"{i+j:4}", end="")
  print()

 100  99  98  97  96  95  94  93  92  91
  90  89  88  87  86  85  84  83  82  81
  80  79  78  77  76  75  74  73  72  71
  70  69  68  67  66  65  64  63  62  61
  60  59  58  57  56  55  54  53  52  51
  50  49  48  47  46  45  44  43  42  41
  40  39  38  37  36  35  34  33  32  31
  30  29  28  27  26  25  24  23  22  21
  20  19  18  17  16  15  14  13  12  11
  10   9   8   7   6   5   4   3   2   1


## Encontrar los números repetidos de una lista

#### Solución 1
Por cada número comparamos con los restantes, para ver si los hay iguales.

In [ ]:
numeros = [3,4,6,2,5,4,2,7,8,0,6,9,1]
print(numeros)
repetidos = []
for i in range(len(numeros)):
  for j in range(i+1,len(numeros)):
    if numeros[i]==numeros[j]:
      repetidos.append(numeros[i])
print("Repetidos: ", repetidos)

[3, 4, 6, 2, 5, 4, 2, 7, 8, 0, 6, 9, 1]
Repetidos:  [4, 6, 2]


#### Solución 2
Recorriendo todos y excluyendo cuando i, j son iguales.

In [ ]:
numeros = [3,4,6,2,5,4,2,7,8,0,6,9,1]
print(numeros)
repetidos = []
for i in range(len(numeros)):
  for j in range(len(numeros)):
    if i!=j and numeros[i]==numeros[j] and numeros[i] not in repetidos:  # la tercera condición se pone para evitar duplicar valores en la lista repetidos
      repetidos.append(numeros[i])
print("Repetidos: ", repetidos)

[3, 4, 6, 2, 5, 4, 2, 7, 8, 0, 6, 9, 1]
Repetidos:  [4, 6, 2]


#### Solución 3
Con un solo ```for```.

In [ ]:
numeros = [3,4,6,2,5,4,2,7,8,0,6,9,1]
print("Números:  ", numeros)
repetidos = []
unicos = []
for n in numeros:
  if n not in unicos:          # si el número no está en la lista unicos
    unicos.append(n)           # se añade a la lista unicos
  else:                        # en caso contrario
    repetidos.append(n)        # se trata de un número repetido y se añade a la lista repetidos
print("Repetidos:", repetidos)
print("Únicos:   ", unicos)

Números:   [3, 4, 6, 2, 5, 4, 2, 7, 8, 0, 6, 9, 1]
Repetidos: [4, 2, 6]
Únicos:    [3, 4, 6, 2, 5, 7, 8, 0, 9, 1]


### Eficiencia si usamos un único bucle ```for```
Vamos a comparar las dos soluciones anteriores midiendo tiempos.

Comprobaremos que la solución más eficiente es la que usa un único bucle ```for```.

In [ ]:
import random
import time
numeros = []
for i in range(10_000):                # creamos una lista que contien 10.000 números
  n = random.randint(1, 10_000_000)    # los números son aleatorios entre 1 y 10 millones
  numeros.append(n)                    # es posible que en la lista existan números repetidos

t0 = time.time()                       # ponemos en marcha el cronómetro

# Solución no eficiente

repetidos = []
for i in range(len(numeros)):
  for j in range(len(numeros)):
    if i!=j and numeros[i]==numeros[j] and numeros[i] not in repetidos:
      repetidos.append(numeros[i])

t1 = time.time()
print("Repetidos:", repetidos)
print(f"Tiempo: {t1-t0} segundos.")
print("="*40)
# Solución eficiente

repetidos = []
unicos = []

t2 = time.time()

for n in numeros:
  if n not in unicos:
    unicos.append(n)
  else:
    repetidos.append(n)

t3 =time.time()
print("Repetidos:", repetidos)
print(f"Tiempo: {t3-t2} segundos.")   # es mucho más eficiente este método

Repetidos: [64289, 3630261, 5855429]
Tiempo: 25.705004930496216 segundos.
Repetidos: [64289, 3630261, 5855429]
Tiempo: 0.9574470520019531 segundos.


### Comprobar si la suma de dos números de una lista es igual al *target*
* Disponemos de una lista de 12 números **con repetición**, entre -9 y +9.
* Comprobar si la suma de dos de ellos da el valor objetivo, denominado *target*.
* *Target* = 4
* Mostrar todas las parejas de números que sumadas den el *target*.
* En la pareja no puede intervenir la suma de un número con sí mismo.
* Pero si el 2 está repetido y el target es 4, si encontraremos la pareja [2,2]
* Si están repetidos tanto el -3 como el 7, solo encontraremos una pareja [-3,7] que sumados den 4.
* Con la lista
 - numeros = [-3, 1, -7, -3, 2, 7, -5, 3, 2, 1, 7, 9]   # están repetidos: -3, 1, 7, 2
* El resultado de las parejas que suman 4 es:
 - [[-3, 7], [1, 3], [2, 2], [-5, 9]]

#### Solución 1

In [2]:
numeros = [-3, 1, -7, -3, 2, 7, -5, 3, 2, 1, 7, 9]   # están repetidos: -3, 1, 7, 2
print(numeros)
parejas = []
target = 4
for i in range(len(numeros)):                      # por cada número de la lista
  for j in range(len(numeros)):                    # recorremos todos los números
    if numeros[i] + numeros[j] == target and i!=j: # si la suma de una pareja es igual al target y el index i es distinto de j
      pareja = sorted([numeros[i], numeros[j]])    # al estar ordenada la pareja
      if pareja not in parejas:                    # podemos comparar la pareja con otra previa
        parejas.append(pareja)                     # de forma que [1,3] no sea distinto de [3,1]
print(parejas)

[-3, 1, -7, -3, 2, 7, -5, 3, 2, 1, 7, 9]
[[-3, 7], [1, 3], [2, 2], [-5, 9]]


#### Solución 2
En este caso en lugar de pedir que i sea distinto de j
* ```i != j```

lo que se piede es que el index j sea mayor que el index i
* ```j > i```

Esto mejora la eficiencia ya que no solo se evitan los casos en los que i y j coinciden sino que en el condicional el filtro exige que j sea estrictamente mayor que i.  

Ejemplo: Supongamos la lista [0,1,2,3,4,5] y el target=5.  
Las parejas serían [[0,5], [1,4], [2,3]]

Valores de i y j que cumplen que j>i son:  
* i=0, j=1
* i=0, j=2
* i=0, j=3
* i=0, j=4
* i=0, j=5    pareja 0+5 = target
* ========
* i=1, j=2
* i=1, j=3
* i=1, j=4    pareja 1+4 = target
* i=1, j=5
* ========
* i=2, j=3    pareja 2+3 = target
* i=2, j=4
* i=2, j=5
* ========
* i=3, j=4
* i=3, j=5
* ========
* i=4, j=5

In [5]:
numeros = [-3, 1, -7, -3, 2, 7, -5, 3, 2, 1, 7, 9]   # están repetidos: -3, 1, 7, 2
print(numeros)
parejas = []
target = 4
for i in range(len(numeros)):
  for j in range(len(numeros)):
    if numeros[i] + numeros[j] == target and j>i:  # si la suma de una pareja es igual al target y el index j es mayor que el i
      pareja = sorted([numeros[i], numeros[j]])    # al estar ordenada la pareja
      if pareja not in parejas:                    # podemos comparar la pareja con otra previa
        parejas.append(pareja)                     # de forma que [1,3] no sea distinto de [3,1]
print(parejas)

[-3, 1, -7, -3, 2, 7, -5, 3, 2, 1, 7, 9]
[[-3, 7], [1, 3], [2, 2], [-5, 9]]


#### Solución 3
* La variable i recorre todos index salvo el último.
* La variable j recorre todos los index desde el i+1
* Esto es más eficiente ya que no es necesario recorrer todos.
* Además, como no es necesario recorrer todos, no es necesario poner la salvedad i!=j, ya que i y j nunca coinciden

In [4]:
numeros = [-3, 1, -7, -3, 2, 7, -5, 3, 2, 1, 7, 9]   # están repetidos: -3, 1, 7, 2
print(numeros)
parejas = []
target = 4
for i in range(len(numeros)-1):       # hasta el penúltimo elemento
  for j in range(i+1, len(numeros)):  # desde el i+1, así se recorren todas las parejas sin repetir
    if numeros[i] + numeros[j] == target:
      pareja = sorted([numeros[i], numeros[j]])
      if pareja not in parejas:
        parejas.append(pareja)
print(parejas)

[-3, 1, -7, -3, 2, 7, -5, 3, 2, 1, 7, 9]
[[-3, 7], [1, 3], [2, 2], [-5, 9]]


#### Solución 4
* Eliminando un bucle ```for``` anidado ganamos en eficiencia.
* La variable t es el target.
* Recorremos todos los valores de la lista números con el for que usa como variable la n.
* Supongamos que el primer valor de la lista números es 3 y que el target es t=4.
* Entonces, se ira buscando por toda la lista el valor t-n = 4-3 = 1.
* Si existe un 1 en la lista números ya hemos encontrado una pareja cuya suma da el target.
* La pareja encontrada es [n, t-n] = [3, 1]
* Con cada loop del bucle ```for``` la variable n va tomando todos los valores de la lista números.
* De esta forma se irań buscando todas las parejas.

In [6]:
numeros = [-3, 1, -7, -3, 2, 7, -5, 3, 2, 1, 7, 9]   # están repetidos: -3, 1, 7, 2
print(numeros)
parejas = []
t = 4         # target
for n in numeros:               # recorremos todos los valores de la lista números
  if t-n in numeros:            # Si t-n está en la lista números, es que ya hemos encontrado una pareja
    pareja = sorted([n, t-n])
    if pareja not in parejas:
      parejas.append(pareja)
print(parejas)

[-3, 1, -7, -3, 2, 7, -5, 3, 2, 1, 7, 9]
[[-3, 7], [1, 3], [2, 2], [-5, 9]]


#### Solución 5
Haciendo una copia

In [7]:
numeros = [-3, 1, -7, -3, 2, 7, -5, 3, 2, 1, 7, 9]   # están repetidos: -3, 1, 7, 2
print(numeros)
parejas = []
t = 4         # target
for n in numeros:
  copia = list(numeros)
  copia.remove(n)
  if t-n in copia:
    pareja = sorted([n, t-n])
    if pareja not in parejas:
      parejas.append(pareja)
print(parejas)

[-3, 1, -7, -3, 2, 7, -5, 3, 2, 1, 7, 9]
[[-3, 7], [1, 3], [2, 2], [-5, 9]]
